[DR-1293] Implement Google Safe Browsing to detect dangerous links on local database

alegarciamina opened the pull request:
## Overview

### Periodically get data for all the lists: _SOCIAL_ENGENEERING_, _MALWARE_ and _UNWANTED_SOFTWARE_.

For each list, the api client provides a _base64-encoded_ string, where the hashes (in binary format), are concatenated into one long string.
Hashes could have different size so this information is also coming from client. 4-bytes long is the more frequent.
It also provides a `State` value indicating the current version of the list to allow partial updates and a _SHA256_ checksum value to ensure local database was correctly updated.

On each request, a `MinimumWaitDuration` value could be provided which indicates the amount of time to wait until the next request. If not provided, there is no restriction.

If the API responds with any _Status Code_ different that 200, the client enters in _Back-off mode_ (until next 200 status code) which means the wait time between requests must be calculated based on a specific formula.

### Check url against local in-memory database

Urls needs to be first canonicalized and then, up to 30 pairs off prefix/suffix expressions must be created.
Each suffix/prefix expression consists of a host suffix (or full host) and a path prefix (or full path).

Each expression will be hashed and check against the Suspicious Lists. If at least one prefix was found, we consider the url _unsafe_.

We are currently checking first against _SOCIAL_ENGENEERING_, then _MALWARE_ and finally against _UNWANTED_SOFTWARE_ lists and creating the `SafeBrowsingResult` with the corresponding `ThreatType` value.
The same url may appear on multiple lists at a time, so we may need to allow to have multiple `ThreatType` in the future.

alegarciamina committed:
[DR-1293] Download google safe browsing initial data

andresmoschini commented on Src/MakingSense.SafeBrowsing/Internal/HttpClient.cs:
```diff
@@ -1,5 +1,4 @@
-﻿#if (!NETSTANDARD1_0)
-using System;
+﻿using System;
```
I think that we could remove this class

andresmoschini replied:
Mmm, maybe not, you are removing the other one, I do not know

andresmoschini commented on Src/MakingSense.SafeBrowsing/Internal/TimerAlternative.cs:
```diff
@@ -23,7 +23,6 @@
 // OTHER DEALINGS IN THE SOFTWARE.
 #endregion
 
-#if (!NET35 && !NET40)
```
I think that this class is not being used and can be removed.

andresmoschini commented:
Sorry, @alegarciamina, at a glance, it looks tidy and nice, but without a deeper reading I am not capable to understand what is it doing and what it is not. Could you add a little description of the scope of this change and how we could take advantage of it? That will help me to read the code easier.

Also, remember to increment nuget package version.

Thanks!

alegarciamina committed:
[DR-1293] Add XML comments to document Google Safe Browsing implementation

alegarciamina committed:
[DR-1293] Ordering SuspiciousLists to support partial updates

andresmoschini commented on Src/MakingSense.SafeBrowsing/GoogleSafeBrowsing/GoogleSafeBrowsingUpdater.cs:
```diff
@@ -0,0 +1,169 @@
+            {
+                if(listUpdate.ResponseType == FULL_UPDATE)
+                {
+                    _database.SuspiciousLists[listUpdate.ThreatType].Hashes = new List<byte[]>();
```
Could yo replace Hashes reference to the new List after create it? in that way it would work better in a multithreath context.

andresmoschini commented on Src/MakingSense.SafeBrowsing/GoogleSafeBrowsing/GoogleSafeBrowsingUpdater.cs:
```diff
@@ -0,0 +1,141 @@
+        /// <param name="clientId">Unique client ID</param>
+        /// <param name="clientVersion">Client version</param>
+        /// <param name="region">Geographic location in ISO 3166-1 alpha-2 format</param>
+        public GoogleSafeBrowsingUpdater(string apiKey, string clientId, string clientVersion, string region = null)
```
What do you think about to receive a configuration object in place of primitives?

andresmoschini commented on Src/MakingSense.SafeBrowsing/GoogleSafeBrowsing/GoogleSafeBrowsingUpdater.cs:
```diff
@@ -0,0 +1,141 @@
+                }
+                else
+                {
+                    //TODO: implement partial updates
```
What do you think about to throw a _not implemented excepction_ here?

andresmoschini commented on Src/MakingSense.SafeBrowsing/GoogleSafeBrowsing/GoogleSafeBrowsingUpdater.cs:
```diff
@@ -0,0 +1,141 @@
+                        _database.SuspiciousLists[listUpdate.ThreatType].Hashes.AddRange(SplitByteList(hashValues, rawHashes.PrefixSize.Value));
+                    }
+
+                    //_database.UnsafeLists[listUpdate.ThreatType].Hashes.Sort();
```
Could you add a clarification about why is this line commented?

andresmoschini commented on Src/MakingSense.SafeBrowsing/GoogleSafeBrowsing/GoogleSafeBrowsingUpdater.cs:
```diff
@@ -0,0 +1,141 @@
+            _database.MinimumWaitDuration = TimeSpan.FromSeconds(double.Parse((response.MinimumWaitDuration as string).TrimEnd('s')));
+            _database.Updated = DateTimeOffset.Now;
+
+            foreach (var listUpdate in response.ListUpdateResponses)
```
I think that this block could be easier to read if it is expressed as Linq expression.

andresmoschini commented on Src/MakingSense.SafeBrowsing/GoogleSafeBrowsing/GoogleSafeBrowsingUpdater.cs:
```diff
@@ -0,0 +1,141 @@
+            {
+                aux.Add(bytes[i]);
+
+                if (aux.Count == size)
```
Should there be a more efficient way to split an array of bytes, maybe indicating _from_ and _to_ values. Maybe it is not the focus right now, in that case, please simply add a _// TODO_ comment indicating that.

andresmoschini commented on Src/MakingSense.SafeBrowsing/MakingSense.SafeBrowsing.csproj:
```diff
@@ -1,6 +1,6 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFrameworks Condition="'$(LibraryFrameworks)'==''">net45;net40;net35;netstandard1.0;netstandard1.3</TargetFrameworks>
+    <TargetFrameworks Condition="'$(LibraryFrameworks)'==''">net45;netstandard1.3</TargetFrameworks>
```
I have said you to remove support for older frameworks. But, then I have remembered that we want to use this library in Doppler (I do not remember what .NET version it is being used there).

I know that the new dependency is not supported by older frameworks, but maybe in the future, we could add a workaround for them. 

So, what do you think about keeping support for older frameworks but do not include the new classes in that packages? 

I can easily create a commit restoring old framework support if you prefer.

alegarciamina committed:
[DR-1293] Code enhancements

alegarciamina committed:
[DR-1293] Partial updates

andresmoschini commented:
Hi @alegarciamina, I have created a new branch keeping the rest of frameworks, see it in my fork: https://github.com/andresmoschini/safe-browsing/commits/google-api, you can rebase your pending work on it.

Also, to run the test, you can execute `.\Build\runbuild.ps1`

And, do not forget to increment the version.

alegarciamina committed:
[DR-1293] Increment package version

andresmoschini commented on Src/MakingSense.SafeBrowsing/GoogleSafeBrowsing/GoogleSafeBrowsingUpdater.cs:
```diff
@@ -0,0 +1,171 @@
+                return;
+            }
+
+            var request = new FetchThreatListUpdatesRequest
```
I am not sure how this google generated API works.  Do you know if it takes advantage of _ETag_ field?

alegarciamina replied:
I don't think so - it comes null for list update requests.

andresmoschini replied:
Thanks!

andresmoschini commented on Src/MakingSense.SafeBrowsing/GoogleSafeBrowsing/GoogleSafeBrowsingUpdater.cs:
```diff
@@ -0,0 +1,171 @@
+                    {
+                        var indices = listUpdate.Removals.SelectMany(x => x.RawIndices.Indices);
+
+                        hashes = hashes.Where((x, index) => !indices.Contains(index));
```
Are you sure that we are making honor to the server indices?

alegarciamina replied:
Yes, we are keeping the lists lexicographically-sorted to be able to correctly process the removal indices.

Anyway, it is pending to implement the Checksum validation to ensure the whole list is correct.

andresmoschini replied:
Great!

alegarciamina committed:
[DR-1293] Enter back-off mode after unsuccessful response

andresmoschini commented:
Do not forget to add some tests.

Maybe you can inject _SafebrowsingService_ instance to _GoogleSafeBrowsingUpdater_, since the class is not sealed and the methods are virtual, I think that it will be easy to create mocks of it.

Also, if you add an _Update_ method to _Database_ object, it will be easy to test its behavior without deal with _Google_ classes.

andresmoschini commented:
> since the class is not sealed and the methods are virtual, I think that it will be easy to create mocks of it.

I am seeing that maybe it is not so easy because we are not using the service directly:

```csharp
var response = await _safebrowsingService.ThreatListUpdates.Fetch(request).ExecuteAsync();
```

Maybe you can create a wrapper for this, something like:

```csharp

interface ISafeBrowsingService
{
    Task<FetchThreatListUpdatesResponse> FetchThreatListUpdatesAsync(FetchThreatListUpdatesRequest request);
}

```

alegarciamina committed:
[DR-1293] Make back-off mode properties read only and exit back-off mode after successful response

alegarciamina committed:
[DR-1293] Encapsulate database update process inside GoogleSafeBrowsingDatabase

alegarciamina committed:
[DR-1293] Wrap Google SafebrowsingService

andresmoschini commented on Src/MakingSense.SafeBrowsing/GoogleSafeBrowsing/GoogleSafeBrowsingUpdater.cs:
```diff
@@ -89,60 +89,23 @@ public GoogleSafeBrowsingUpdater(GoogleSafeBrowsingConfiguration configuration)
-            _database.MinimumWaitDuration = TimeSpan.FromSeconds(double.Parse((response.MinimumWaitDuration as string).TrimEnd('s')));
-            _database.Updated = DateTimeOffset.Now;
+            var minWaitDuration = TimeSpan.FromSeconds(double.Parse((response.MinimumWaitDuration as string).TrimEnd('s')));
+            var listUpdates = response.ListUpdateResponses.Select(x=> MapListUpdate(x));
```
If you like, you can rewrite it as `var listUpdates = response.ListUpdateResponses.Select(MapListUpdate);`

alegarciamina committed:
[OT] Remove dotnet.exe --no-build argument to avoid appveyor error

See https://github.com/dotnet/BenchmarkDotNet/issues/530

alegarciamina committed:
[DR-1293] Fix error when MinimumWaitDuration comes null

alegarciamina committed:
[DR-1293] Tests

alegarciamina committed:
[DR-1293] Helper to generate hashes

alegarciamina committed:
[DR-1293] Create canonical urls

alegarciamina committed:
[DR-1293] Create suffixes/prefixes patterns

alegarciamina committed:
[DR-1293] Check url in local database

alegarciamina committed:
[DR-1293] Validate Checksum

alegarciamina committed:
[DR-1293] GoogleSafeBrowsing Checker tests

alegarciamina committed:
[DR-1293] Improving FindPrefixes performance by using BinarySearch

alegarciamina committed:
[DR-1293] Improving Update by avoiding unnecessary ordering.

Hashes are coming lexicography ordered from API, there is no need to order them.

alegarciamina committed:
[DR-1293] Periodically serialize database and save to json file

andresmoschini commented on Src/MakingSense.SafeBrowsing.Tests/CanonicalUrlTests.cs:
```diff
@@ -0,0 +1,137 @@
+        [TestCase("  http://www.google.com/  ", "http://www.google.com/")]
+        [TestCase("http:// leadingspace.com/", "http://%20leadingspace.com/")]
+        [TestCase("http://%20leadingspace.com/", "http://%20leadingspace.com/")]
+        [TestCase("%20leadingspace.com/", "http://%20leadingspace.com/")]
```
Is this fine? should leading space not be trimmed?

alegarciamina replied:
Yes, these test cases are provided by Google [here](https://developers.google.com/safe-browsing/v4/urls-hashing#canonicalization) to test canonicalization.

andresmoschini commented on Src/MakingSense.SafeBrowsing/CanonicalUrl.cs:
```diff
@@ -0,0 +1,233 @@
+
+namespace MakingSense.SafeBrowsing
+{
+    public class CanonicalUrl
```
Nice class! I wonder if there is already a C# out of the box class with some of this functionality. Anyway, it looks very clear. Did you copy the code from another language implementation or do you have a link to the specification?

alegarciamina replied:
The **canonicalization** part was taken from [here](https://searchcode.com/codesearch/view/17874600/). I've only applied some changes to regular expressions to meet [specifications](https://developers.google.com/safe-browsing/v4/urls-hashing#canonicalization) and pass all the suggested test cases, and some minor code improvement.

I've tried to replace Unescape method with Uri.UnescapeDataString, but this method fails for NET3.5 with some urls that are really strange like http:// leadingspace.com/ but included in recommended test cases.

Regarding **suffix/prefix patterns**, it's a custom implementation based on [specifications](https://developers.google.com/safe-browsing/v4/urls-hashing#suffixprefix-expressions).

andresmoschini commented:
Great work @alegarciamina! :shipit:

andresmoschini commented on Src/MakingSense.SafeBrowsing/GoogleSafeBrowsing/GoogleSafeBrowsingDatabase.cs:
```diff
@@ -174,18 +179,33 @@ public void Update(DateTimeOffset now, TimeSpan? minWaitDuration, IEnumerable<Li
             Updated = now;
         }
 
+        public static GoogleSafeBrowsingDatabase Initialize(string snapshotPath = null)
```
Great! are you even writing and reading the database from filesystem!

alegarciamina commented:
Sample call to check a suspicious prefix against API ([fullHashes.find](https://developers.google.com/safe-browsing/v4/update-api#example-fullHashesfind)):

![image](https://user-images.githubusercontent.com/25298072/35233548-379acdc2-ff7d-11e7-84d9-1ccfdc5ec823.png)
